src/Microsoft.OData.Core/UriParser/Resolver/NormalizedModelElementsCache.cs

```diff
@@ -89,26 +93,76 @@ public List<IEdmTerm> FindTerms(string qualifiedName)
             return null;
         }
 
+        /// <summary>
+        /// Find all navigation sources that match the <paramref name="name"/>.
+        /// </summary>
+        /// <param name="name">The case-insensitive name to match.</param>
+        /// <returns>A list of matching navigation sources, or null if no navigation source matches the name.</returns>
+        public List<IEdmNavigationSource> FindNavigationSources(string name)
+        {
+            if (navigationSourcesCache.TryGetValue(name, out List<IEdmNavigationSource> results))

```

Comment:
Is the name the simple identifier of the navigation source? not the qualified name?

How about searching the navigation source in the referenced model? with the same name but a different namespace?

Reply:
Yes, it's the simple identifier of the navigation source. And yes, it's only searching the top-level model entity container.

The reason I implemented it this way is because I wanted to retain the same behaviour as the existing implementation which performs the search as follows:

```c#
IEdmEntityContainer container = model.EntityContainer;
if (container == null)
{
    return null;
}

var result = container.Elements.OfType<IEdmNavigationSource>()
    .Where(source => string.Equals(identifier, source.Name, StringComparison.OrdinalIgnoreCase));
```
see: https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs#L94-L102

As you can see the existing implementation only searches the top-level model's container and it only compares the `source.Name`, not the fully qualified name (including the container name).

If I change the behaviour of the cache to include referenced models or fully qualified identifier, then there would be an inconsistency between the cache and non-cache code paths.

On the other hand, if the expected behaviour is that the URI resolver should search referenced models and/or fully qualified name, then the current implementation is a bug. And I think that should be a separate discussion since that change of behaviour could be observable to the customer.

Reply:
The [doc comment](https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs#L283) of the `ResolveOperationImports` describes the `identifier` parameter as follows:
```
The qualified name of the operation import which may or may not include the container name.
```

However, the existing implementation does not consider the qualified name in its search:

```c#
IEdmEntityContainer container = model.EntityContainer;
if (container == null)
{
    return Enumerable.Empty<IEdmOperationImport>();
}

return container.Elements.OfType<IEdmOperationImport>()
    .Where(source => string.Equals(identifier, source.Name, StringComparison.OrdinalIgnoreCase));
```

Reply:
If we are specifying the navigation source identifier and not searching in referenced models, is there a scenario where whatever that gets returned is a list and not just a single navigation source?

Reply:
Like can we have several navigation sources with the same identifier in the same model?

Reply:
Since we're searching a single container, I don't think that's possible. Maybe if we were searching in multiple containers that could be possible, but that would be arguably a new feature, extending the existing behaviour, which is beyond the scope of this PR.

Reply:
However, because we are dealing with a case insensitive scenario, we can have multiple navigation sources in the same container whose names only differ by case. That's why we return a list.

Reply:
can we have navigation sources in the same container whose names only differ by case? or we can access a navigation source using different cases?

Reply:
The edm model is case-sensitive by default, so it is possible to have navigation sources which only differ by case because they are different names as far as the `IEdmModel` is concerned. If the `IEdmModel` were case-insensitive, then this cache would not be necessary.

Since this cache is case-insensitive, for the same key we might have multiple matching entries, that's why were store the values in lists.